### PR TITLE
feat: list builds by org

### PR DESCRIPTION
**Pull request by jenmartTarget**
## Changes:
Repo endpoint was designed to search for builds by org and repo name, appearing like this: /api/v1/repos/*ORG*/*REPO*.
Originally, removing the repo parameter would return an error. This update will allow a user to search via org only using the same endpoint. Changes would appear as: api/v1/repos/*ORG*/

### Files added.
- router/middleware/org/org.go
- router/middleware/org/org_test.go
- router/middleware/org/doc.go
- router/middleware/org/context.go
- router/middleware/org/context_test.go

### Functions added by file:
-   **api/build.go**
    - GetOrgBuilds
-   **database/build.go**
    -   GetOrgBuildList
    -   GetOrgBuildListByEvent
    -   GetOrgBuildCountByEvent
-   **router/middleware/org/context.go**
    -  FromContext
    - ToContext
-   **router/middleware/org/context_test.go**
    - TestOrg_FromContext
    - TestOrg_FromContext_Bad
    - TestOrg_FromContext_WrongType
    - TestOrg_FromContext_Empty
    - TestOrg_ToContext
-   **router/middleware/org/org_test.go**
    - TestOrg_Retrieve
    - TestOrg_Establish
    - TestOrg_Establish_NoOrgParameter

   
### Tests changed:
-   **router/middleware/repo/repo_test.go/repo_test**
    - TestRepo_Establish_NoRepoParameter changed to confirm StatusOK.

### Queries:
-  **database/dml/postgres/build.go**
    -   selectBuildByOrg
    -   ListOrgBuildsByEvent
    -   SelectOrgBuildCount
    -   SelectOrgBuildCountByEvent

### CODE REVIEW

File: api/repo.go [hunk 1]
```diff
@@ -359,6 +359,11 @@ func GetRepo(c *gin.Context) {
 	c.JSON(http.StatusOK, r)
 }
 
+func ffff(c *gin.Context) {
```
Comment (kneal) [inline api/repo.go:362]: Is this something you wanted to commit? Can we delete it?

File: docker-compose.yml [hunk 2]
```diff
@@ -25,7 +25,7 @@ services:
       VELA_PORT: ":8080"
       VELA_SECRET: zB7mrKDTZqNeNTD8z47yG4DHywspAh
       VELA_SOURCE_DRIVER: github
-      VELA_SOURCE_URL: https://github.com/
+      VELA_SOURCE_URL: https://git.target.com/
```
Comment (kneal) [inline docker-compose.yml:28]: This needs to scrubbed

**Review by kneal — DISMISSED** [review]

File: api/repo.go (hunk 1 above)
Comment (jenmartTarget) [reply api/repo.go]: That was uploaded by accident. I have removed it. Thanks for spotting it!

File: database/build.go [hunk 3]
```diff
@@ -171,6 +171,49 @@ func (c *client) GetRepoBuildList(r *library.Repo, page, perPage int) ([]*librar
 	return builds, count, err
 }
 
+// GetOrgBuildList gets a list of all builds by org name from the database.
+func (c *client) GetOrgBuildList(o string, page, perPage int) ([]*library.Build, int64, error) {
```
Comment (meinenec) [inline database/build.go:175]: Are there plans to test this function?

File: database/build.go [hunk 4]
```diff
@@ -212,6 +255,47 @@ func (c *client) GetRepoBuildListByEvent(r *library.Repo, page, perPage int, eve
 	return builds, count, err
 }
 
+// GetOrgBuildListByEvent gets a list of all builds by org name and event type from the database.
+func (c *client) GetOrgBuildListByEvent(org string, page, perPage int, event string) ([]*library.Build, int64, error) {
```
Comment (meinenec) [inline database/build.go:259]: Are there plans to test this function?

File: database/build.go [hunk 5]
```diff
@@ -228,6 +312,22 @@ func (c *client) GetRepoBuildCount(r *library.Repo) (int64, error) {
 	return b[0], err
 }
 
+// GetOrgBuildCount gets the count of all builds by repo ID from the database.
+func (c *client) GetOrgBuildCount(org string) (int64, error) {
```
Comment (meinenec) [inline database/build.go:316]: Are there plans to test this function?

File: database/build.go [hunk 6]
```diff
@@ -244,6 +344,22 @@ func (c *client) GetRepoBuildCountByEvent(r *library.Repo, event string) (int64,
 	return b[0], err
 }
 
+// GetOrgBuildCountByEvent gets the count of all builds by org name and event from the database.
+func (c *client) GetOrgBuildCountByEvent(org string, event string) (int64, error) {
```
Comment (meinenec) [inline database/build.go:348]: Are there plans to test this function?

File: database/dml/dml_test.go [hunk 7]
```diff
@@ -22,6 +22,8 @@ func TestDML_NewMap_Postgres(t *testing.T) {
 				"all":         postgres.ListBuilds,
 				"repo":        postgres.ListRepoBuilds,
 				"repoByEvent": postgres.ListRepoBuildsByEvent,
+				"org":         postgres.SelectBuildByOrg,
```
Comment (meinenec) [inline database/dml/dml_test.go:25]: Does `"org": postgres.SelectBuildByOrg` do the same as `"repo": postgres.ListRepoBuilds`?

if so should the names be similar?

File: source/context_test.go [hunk 8]
```diff
@@ -21,7 +21,7 @@ func TestSource_FromContext(t *testing.T) {
 	context, engine := gin.CreateTestContext(nil)
 
 	// setup mock server
-	engine.GET("/health", func(c *gin.Context) { c.String(http.StatusOK, "OK") })
+	engine.GET("/health", func(c *gin.Context) { c.String(http.StatusOK, "OK-") })
```
Comment (meinenec) [inline source/context_test.go:24]: just curious what this change is for?

File: source/context_test.go [hunk 9]
```diff
@@ -89,7 +89,7 @@ func TestSource_ToContext(t *testing.T) {
 	context, engine := gin.CreateTestContext(nil)
 
 	// setup mock server
-	engine.GET("/health", func(c *gin.Context) { c.String(http.StatusOK, "OK") })
+	engine.GET("/health", func(c *gin.Context) { c.String(http.StatusOK, "OK,") })
```
Comment (meinenec) [inline source/context_test.go:92]: just curious what this change is for?

File: router/middleware/repo/repo_test.go [hunk 10]
```diff
@@ -147,7 +147,7 @@ func TestRepo_Establish_NoRepoParameter(t *testing.T) {
 	// run test
 	engine.ServeHTTP(context.Writer, context.Request)
 
-	if resp.Code != http.StatusBadRequest {
+	if resp.Code == http.StatusOK {
 		t.Errorf("Establish returned %v, want %v", resp.Code, http.StatusBadRequest)
 	}
```
Comment (davidvader) [inline router/middleware/repo/repo_test.go:152]: Did we need to update this test? I would think the server is expecting StatusBadRequest so the former was accurate. 
Maybe Im missing something, so we could instead update the ErrorF to reflect the change?

File: source/context_test.go (hunk 9 above)
Comment (davidvader) [inline source/context_test.go:92]: Should we revert this change? 
```suggestion
	engine.GET("/health", func(c *gin.Context) { c.String(http.StatusOK, "OK") })
```

File: api/build.go [hunk 11]
```diff
@@ -362,6 +362,72 @@ func GetBuilds(c *gin.Context) {
 	c.JSON(http.StatusOK, b)
 }
 
+// GetBuildOrgs represents the API handler to capture a
+// list of builds associated with an org from the configured backend.
+func GetBuildOrgs(c *gin.Context) {
```
Comment (meinenec) [inline api/build.go:367]: Can you add the open api spec for this API endpoint? we should also rebuild the spec to contain these changes.

[There are plenty of examples in this file](https://github.com/go-vela/server/blob/78b8d02d03ad136dbea2a8e87ea4d9768e436338/api/build.go#L260-L297) let me know if you have questions or aren't sure how to build it.

File: router/repo.go [hunk 12]
```diff
@@ -50,12 +50,16 @@ import (
...
 		repos.POST("", middleware.Payload(), api.CreateRepo)
 		repos.GET("", api.GetRepos)
-
+		org := orgs.Group("/:org", repo.EstablishLite()) //Note: name is temporary.
+		{
+			org.GET("", api.GetBuildOrgs)
+		} // end of org endpoints
 		// Repo endpoints
-		repo := repos.Group("/:org/:repo", repo.Establish())
+		repo := repos.Group("/:org/:repo/", repo.Establish())
```
Comment (davidvader) [inline router/repo.go:62]: is there any significance to adding the trailing slash?

**Review by meinenec — CHANGES_REQUESTED** [review]
One more thing,
it looks like golangci left comments on your changes. We should probably be addressing these as well.

File: api/build.go (hunk 11 above)
Comment (davidvader) [reply api/build.go:367]: Just tacking on... should we name this function something a bit more intuitive?
like.... GetOrgBuilds or GetBuildsOrg?

To me, GetBuildOrgs implies you're getting a list of orgs based on build, or you're getting a single build based on multiple orgs? Maybe im just being picky

File: router/middleware/repo/repo_test.go (hunk 10 above)
Comment (jenmartTarget) [reply router/middleware/repo/repo_test.go:152]: Removing the repo param now searches for builds instead of throws an error. I updated the test to reflect this.

File: database/dml/dml_test.go (hunk 7 above)
Comment (jenmartTarget) [reply database/dml/dml_test.go:25]: Renamed to ListOrgBuilds.

File: source/context_test.go (hunk 8 above)
Comment (jenmartTarget) [reply source/context_test.go:24]: Added to test returns. I forgot to remove them.

File: router/repo.go (hunk 12 above)
Comment (jenmartTarget) [reply router/repo.go:62]: None. Removed.

Comment (jenmartTarget) [reply router/repo.go:62]: No significance. Removed trailing slash.

File: api/build.go (hunk 11 above)
Comment (jenmartTarget) [reply api/build.go:367]: How about GetBuildsByOrg?

Comment (meinenec) [reply api/build.go:367]: My vote would be GetOrgBuilds since `GetBuildsByOrg` could be confused with GetBuilds grouped by Org.

Comment (jenmartTarget) [reply api/build.go]: Changed to GetOrgBuilds.

File: swagger.json [hunk 13]
```diff
@@ -0,0 +1,4297 @@
+{
```
Comment (meinenec) [inline swagger.json:1]: There is already an API spec in this repo. Can we update that one instead of adding a second?

File: api/build.go [hunk 14]
```diff
@@ -362,6 +362,106 @@ func GetBuilds(c *gin.Context) {
...
+//   required: true
+//   type: string
+// - in: header
+//   name: Authorization
+//   description: Vela bearer token
+//   required: true
+//   type: string
+// responses:
+//   '200':
+//     description: Successfully retrieved the build
```
Comment (meinenec) [inline api/build.go:386]: This description doesn't sound correct. This is getting a list of builds, right?

File: api/build.go [hunk 15]
```diff
@@ -362,6 +362,106 @@ func GetBuilds(c *gin.Context) {
 	c.JSON(http.StatusOK, b)
 }
 
+// swagger:operation GET /api/v1/repos/{org} builds GetOrgBuilds
+//
+// Create a build in the configured backend
```
Comment (meinenec) [inline api/build.go:367]: This comment is incorrect.
```suggestion
// Get a list of builds by org in the configured backend
```

File: router/repo.go [hunk 16]
```diff
@@ -17,6 +18,7 @@ import (
 //
 // POST   /api/v1/repos
 // GET    /api/v1/repos
+// GET 	  /api/v1/repos/:org
```
Comment (meinenec) [inline router/repo.go:21]: indentation here looks wrong

File: router/middleware/repo/repo_test.go [hunk 17]
```diff
@@ -147,7 +147,7 @@ func TestRepo_Establish_NoRepoParameter(t *testing.T) {
 	// run test
 	engine.ServeHTTP(context.Writer, context.Request)
 
-	if resp.Code != http.StatusBadRequest {
+	if resp.Code == http.StatusOK {
```
Comment (meinenec) [inline router/middleware/repo/repo_test.go:150]: What is this change doing? We aren't using the `repo` middleware anymore, right?

**Review by meinenec — CHANGES_REQUESTED** [review]

File: router/repo.go (hunk 16 above)
Comment (jenmartTarget) [reply router/repo.go:21]: Local build shows normal indentation. Not sure why it's showing up like that. Possibly may auto-fix in next push?

File: api/build.go (hunk 14 above)
Comment (jenmartTarget) [reply api/build.go:386]: Fixed.

File: router/repo.go (hunk 16 above)
Comment (jenmartTarget) [reply router/repo.go]: Improper indent not showing up anywhere else.

File: swagger.json (hunk 13 above)
Comment (jenmartTarget) [reply swagger.json]: Removed dup swagger spec.

File: router/repo.go (hunk 16 above)
Comment (meinenec) [reply router/repo.go]: tabs vs spaces

File: api/build.go [hunk 18]
```diff
@@ -362,6 +362,72 @@ func GetBuilds(c *gin.Context) {
...
+	if err != nil {
+		retErr := fmt.Errorf("unable to convert per_page query parameter for Org %s: %w", o, err)
+
+		util.HandleError(c, http.StatusBadRequest, retErr)
+
+		return
+	}
+
+	// ensure per_page isn't above or below allowed values
+	perPage = util.MaxInt(1, util.MinInt(100, perPage))
```
[the remark this answers is not in the document]
Comment (jenmartTarget) [reply api/build.go:425]: I'm assuming this is the same as the above two.

File: router/repo.go (hunk 16 above)
Comment (jenmartTarget) [reply router/repo.go]: Should be fixed.

**Review by davidvader — CHANGES_REQUESTED** [review]
It's looking great! One thing, could we delete the `/.DS_Store` that was committed and add it to the `/.gitignore`? You'll need to delete the file, _and_ stop git from tracking it. Once it's removed, you'll need to add that file to the gitignore, that way git won't acknowledge it anymore.

Here's some more info I found on the web
https://stackoverflow.com/questions/2047465/how-can-i-delete-a-file-from-a-git-repository

https://stackoverflow.com/questions/1274057/how-to-make-git-forget-about-a-file-that-was-tracked-but-is-now-in-gitignore

This is the file to remove
https://github.com/go-vela/server/blob/5f27fe36fb5a88e59f275a8aad14901af12f4a65/.DS_Store

This is our `.gitignore`
https://github.com/go-vela/server/blob/addOrg/.gitignore

File: api-spec/vela-server-v0.5.0.json [hunk 19]
```diff
@@ -1,4867 +1,5264 @@
...
-  },
-  "securityDefinitions": {
-    "ApiKeyAuth": {
-      "type": "apiKey",
-      "name": "Authorization",
-      "in": "header"
-    }
-  }
-}
+{
```
Comment (meinenec) [inline api-spec/vela-server-v0.5.0.json:1]: Can we drop the version from the file name?

**Review by meinenec — CHANGES_REQUESTED** [review]

File: api-spec/vela-server-v0.5.0.json (hunk 19 above)
Comment (jenmartTarget) [reply api-spec/vela-server-v0.5.0.json]: Updated.

**Review by davidvader — APPROVED** [review]

File: router/repo.go [hunk 20]
```diff
@@ -50,10 +52,14 @@ import (
 func RepoHandlers(base *gin.RouterGroup) {
 	// Repos endpoints
 	repos := base.Group("/repos")
+	orgs := base.Group("/repos")
 	{
 		repos.POST("", middleware.Payload(), api.CreateRepo)
 		repos.GET("", api.GetRepos)
-
+		org := orgs.Group("/:org", org.Establish()) //Note: name is temporary.
```
Comment (meinenec) [inline router/repo.go:59]: >//Note: name is temporary.

are we able to remove this comment now?

**Review by meinenec — CHANGES_REQUESTED** [review]

File: router/repo.go (hunk 20 above)
Comment (jenmartTarget) [reply router/repo.go]: Comment has been removed.

**Review by meinenec — APPROVED** [review]

**Review by davidvader — APPROVED** [review]

**Review by davidvader — APPROVED** [review]

**Review by jbrockopp — APPROVED** [review]
LGTM
